include/libopencm3/sam/d/port.h

```diff
@@ -66,3 +134,216 @@
+
+/* PORTx_IN[31:0]: INy[31:0]: Port input data [y=0..31] */
+
+/* --- PORTx_CTRL values --------------------------------------------------- */
+
+/* PORTx_CTRL[31:0]: CTRLy[31:0]: Port input sampling mode [y=0..31] */
+
+/* --- PORTx_WRCONFIG values ----------------------------------------------- */
+
+/* HWSEL: Half word select: 0 [15:0], 1 [31:16] */
+#define PORT_WRCONFIG_HWSEL         (1 << 31)
+
+/* WRPINCFG: Write PINCFG: 1 to update pins for selected by PINMASK */
+#define PORT_WRCONFIG_WRPINCFG      (1 << 30)
+
+/* Bit 29: Reserved */
+
+/* WRPMUX: Write PMUX: 1 to update pins pmux for selected by PINMASK */
+#define PORT_WRCONFIG_WRPMUX        (1 << 28)
+
+/* PMUX: Peripheral Multiplexing: determine pmux for pins selected by PINMASK */
+#define PORT_WRCONFIG_PMUX(mux)     ((0xf & (mux)) << 24)
+
+/* Bit 23: Reserved */
+
+/* DRVSTR: Output Driver Strength Selection: determine strength for pins in PINMASK */
+#define PORT_WRCONFIG_DRVSTR        (1 << 22)
+
+/* Bit [21:19]: Reserved */
+
+/* PULLEN: Pull Enable: enable PINCFGy.PULLEN for pins in PINMASK */
+#define PORT_WRCONFIG_PULLEN        (1 << 18)
+
+/* INEN: Input Enable: enable PINCFGy.INEN for pins in PINMASK */
+#define PORT_WRCONFIG_INEN          (1 << 17)
+
+/* PMUXEN: Peripheral Multiplexer Enable: enable PINCFGy.PMUXEN for pins in PINMASK */
+#define PORT_WRCONFIG_PMUXEN        (1 << 16)
+
+/* PINMASK: Pin Mask for Multiple Pin Configuration: select pins to be configured
+ *          [31:16] if HWSET=1, [15:0] if HWSET=0
+ */
+#define PORT_WRCONFIG_PINMASK(pins) ((0xffff & (pins)) << 0)
+
+/* --- PORTx_PMUX values --------------------------------------------------- */
+
+/* PMUXO: Peripheral Multiplexing for Odd-Numbered Pin: 2*x+1 pin multiplexing */
+#define PORT_PMUX_PMUXO(mux)        ((0xf & (mux)) << 4)
+
+/* PMUXE: Peripheral Multiplexing for Even-Numbered Pin: 2*x pin multiplexing */
+#define PORT_PMUX_PMUXE(mux)        ((0xf & (mux)) << 0)
+
+/* --- PORTx_PINCFGy values ------------------------------------------------ */
+
+/* Bit 7: Reserved */
+
+/* DRVSTR: Output Driver Strength Selection */
+#define PORT_PINCFG_DRVSTR (1 << 6)
+
+/* Bit [5:3]: Reserved */
+
+/* PULLEN: Pull Enable */
+#define PORT_PINCFG_PULLEN (1 << 2)
+
+/* INEN: Input Enable */
+#define PORT_PINCFG_INEN   (1 << 1)
+
+/* PMUXEN: Peripheral Multiplexer Enable */
+#define PORT_PINCFG_PMUXEN (1 << 0)
+
+/* --- Convenience enums --------------------------------------------------- */
+
+/* GPIO mode definitions (for convenience) */
+/** @defgroup gpio_mode GPIO mode configuration
+@ingroup gpio_mode
+@{*/
+enum gpio_mode {
+	GPIO_MODE_OUTPUT,
+	GPIO_MODE_INOUT,
+	GPIO_MODE_INPUT_FLOAT,
+	GPIO_MODE_INPUT_PULLDOWN,
+	GPIO_MODE_INPUT_PULLUP,
+	GPIO_MODE_AF
+};
+/**@}*/
+
+/* --- Function prototypes ------------------------------------------------- */
+
+BEGIN_DECLS
+
+void gpio_mode_setup(uint32_t gpioport, enum gpio_mode mode, uint32_t gpios);
```

Comment:
most most chips here (except efm32 it seems) are are using mode setup with pull/down parameter :

> void gpio_mode_setup(uint32_t gpioport, uint8_t mode, uint8_t pull_up_down, uint16_t gpios)

maybe we could stick to that, and keep the same gpio_mode enum ?